components/org.wso2.identity.event.websubhub.publisher/src/main/java/org/wso2/identity/event/websubhub/publisher/service/WebSubEventSubscriberImpl.java

```diff
@@ -0,0 +1,195 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.buildSubscriptionURL;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.constructHubTopic;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.getWebSubBaseURL;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleErrorResponse;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleFailedOperation;
+import static org.wso2.identity.event.websubhub.publisher.util.WebSubHubAdapterUtil.handleSuccessfulOperation;
+
+/**
+ * OSGi service for managing WebSub Hub subscriptions.
+ */
+public class WebSubEventSubscriberImpl implements WebhookSubscriber {
+
+    private static final Log log = LogFactory.getLog(WebSubEventSubscriberImpl.class);
+
+    @Override
+    public String getName() {
+
+        return WebSubHubAdapterConstants.WEB_SUB_HUB_ADAPTER_NAME;
+    }
+
+    @Override
+    public boolean subscribe(List<String> topics, String callbackUrl, String tenantDomain) {
+
+        try {
+            // Create a list to hold all the futures
+            List<CompletableFuture<Boolean>> futures = new ArrayList<>();
+
+            for (String topic : topics) {
+                futures.add(CompletableFuture.supplyAsync(() -> {
+                    try {
+                        makeSubscriptionAPICall(constructHubTopic(topic, tenantDomain),
+                                getWebSubBaseURL(),
+                                WebSubHubAdapterConstants.Http.SUBSCRIBE,
+                                callbackUrl);
+                        log.debug("WebSub Hub subscription successful for topic: " + topic +
+                                " with callback URL: " + callbackUrl + " in tenant: " + tenantDomain);
+                        return true;
+                    } catch (WebSubAdapterException e) {
+                        log.error("Error subscribing to topic: " + topic, e);
+                        return false;
+                    }
+                }));
+            }
+
+            // Combine all the futures into a single future and wait for completion
+            CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+            // Wait for all futures to complete
+            allFutures.join();
+
+            // Check if any subscription failed
+            return futures.stream().allMatch(CompletableFuture::join);
+
+        } catch (Exception e) {
+            log.error("Unexpected error during subscription process", e);
+            return false;
+        }
+    }
+
+    @Override
+    public boolean unsubscribe(List<String> topics, String callbackUrl, String tenantDomain)
+            throws WebhookMgtException {
```

Comment:
We can make this exeception defined as WebhookSubscriptionException and it should be thrown from both subscribe and unsubscribe methods default

Reply:
Address via https://github.com/wso2/product-is/issues/24085